src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Arcus.Messaging.Abstractions;
+using GuardNet;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Arcus.Messaging.Pumps.Abstractions
+{
+    public abstract class MessagePump<TMessage, TMessageContext> : BackgroundService
+        where TMessageContext : MessageContext
+    {
+        /// <summary>
+        ///     Logger to write telemetry to
+        /// </summary>
+        protected ILogger Logger { get; }
+
+        /// <summary>
+        ///     Configuration of the application
+        /// </summary>
+        protected IConfiguration Configuration { get; }
+
+
+        /// <summary>
+        ///     Collection of services that are configured
+        /// </summary>
+        protected IServiceProvider ServiceProvider { get; }
+
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="configuration">Configuration of the application</param>
+        /// <param name="serviceProvider">Collection of services that are configured</param>
+        /// <param name="logger">Logger to write telemetry to</param>
+        protected MessagePump(IConfiguration configuration, IServiceProvider serviceProvider, ILogger logger)
+        {
+            Guard.NotNull(logger, nameof(logger));
+            Guard.NotNull(configuration, nameof(configuration));
+            Guard.NotNull(serviceProvider, nameof(serviceProvider));
+
+            Logger = logger;
+            Configuration = configuration;
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        ///     Handles an exception that occured during the receiving of a message
+        /// </summary>
+        /// <param name="receiveException">Exception that occured</param>
+        protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
+        {
+            Logger.LogError(receiveException, "Unable to process message");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///     Deserializes a raw JSON message body
+        /// </summary>
+        /// <param name="rawMessageBody">Raw message body to deserialize</param>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Deserialized message</returns>
+        protected virtual TMessage DeserializeJsonMessageBody(byte[] rawMessageBody, MessageContext messageContext)
+        {
+            Encoding encoding = DetermineMessageEncoding(messageContext);
+            string serializedMessageBody = encoding.GetString(rawMessageBody);
+
+            TMessage messageBody = JsonConvert.DeserializeObject<TMessage>(serializedMessageBody);
```

Comment:
Probably not for this PR, but could be useful to add options to control deserialization. Enum/int, null handling, allowed types, ... all JSON settings.

Reply:
Tracking in https://github.com/arcus-azure/arcus.messaging/issues/43